File: pkg/wifi/event.go
package wifi

import (
	"errors"
	"strings"
)

type EventType string

const (
	EventTypeTempReached EventType = "temp_reached"
	EventTypeLowWater    EventType = "low_water"
	EventTypeStart       EventType = "start"
	EventTypeStop        EventType = "stop"
	EventTypeChangeTemp  EventType = "change_temp"
	EventTypeTimeStart   EventType = "time_start"
	EventTypeTimeStop    EventType = "time_stop"
	EventTypeTimeFinish  EventType = "time_finish"
	EventTypeChangeParam EventType = "change_param"
)

type EventOriginator string

const (
	EventOriginatorWiFi   EventOriginator = "wifi"
	EventOriginatorBLE    EventOriginator = "ble"
	EventOriginatorDevice EventOriginator = "device"
)

type AnovaEvent struct {
	Type       EventType       `json:"type"`
	Originator EventOriginator `json:"originator"`
}

func ParseEvent(msg *AnovaMessage) (AnovaEvent, error) {
	eventString := string(*msg)
	orig := EventOriginatorDevice

	eventString = strings.TrimSpace(strings.ToLower(eventString))

	if strings.HasPrefix(eventString, "event wifi") {
		orig = EventOriginatorWiFi
		eventString = strings.TrimPrefix(eventString, "event wifi ")
	} else if strings.HasPrefix(eventString, "event ble") {
		orig = EventOriginatorBLE
		eventString = strings.TrimPrefix(eventString, "event ble ")
	} else if strings.HasPrefix(eventString, "event ") {
		eventString = strings.TrimPrefix(eventString, "event ")
	}

	var eventType EventType

	switch {
	case strings.HasPrefix(eventString, "user changed"):
		eventType = EventTypeChangeParam
	case eventString == "stop":
		eventType = EventTypeStop
	case eventString == "start":
		eventType = EventTypeStart
	case eventString == "low water":
		eventType = EventTypeLowWater
	case eventString == "time start":
		eventType = EventTypeTimeStart
	case eventString == "time stop":
		eventType = EventTypeTimeStop
	case eventString == "time finish":
		eventType = EventTypeTimeFinish
	case strings.HasPrefix(eventString, "temp has reached"):
		eventType = EventTypeTempReached
	default:
		return AnovaEvent{}, errors.New("unknown event: " + eventString)
	}

	return AnovaEvent{Type: eventType, Originator: orig}, nil
}

func IsEvent(msg *AnovaMessage) bool {
	return strings.HasPrefix(string(*msg), "event") || strings.HasPrefix(string(*msg), "user changed")
}

File: python/src/anova_wifi/event.py
from enum import Enum

from pydantic import BaseModel


class EventType(str, Enum):
    TEMP_REACHED = "temp_reached"
    LOW_WATER = "low_water"
    START = "start"
    STOP = "stop"
    CHANGE_TEMP = "change_temp"
    TIME_START = "time_start"
    TIME_STOP = "time_stop"
    TIME_FINISH = "time_finish"
    ChangeParam = "change_param"


class EventOriginator(str, Enum):
    WIFI = "wifi"
    BLE = "ble"
    Device = "device"


class AnovaEvent(BaseModel):
    type: EventType
    originator: EventOriginator = EventOriginator.Device

    @classmethod
    def parse_event(cls, event_string: str) -> 'AnovaEvent':
        orig = EventOriginator.Device

        if event_string.startswith("event wifi"):
            orig = EventOriginator.WIFI
        elif event_string.startswith("event ble"):
            orig = EventOriginator.BLE

        event_string = event_string.replace("event ", "").replace("wifi ", "").replace("ble ", "")

        es = event_string.lower().strip()
        if es.startswith("user changed"):
            return cls(type=EventType.ChangeParam, originator=orig)
        elif es == "stop":
            return cls(type=EventType.STOP, originator=orig)
        elif es == "start":
            return cls(type=EventType.START, originator=orig)
        elif es == "low water":
            return cls(type=EventType.LOW_WATER, originator=orig)
        elif es == "time start":
            return cls(type=EventType.TIME_START, originator=orig)
        elif es == "time stop":
            return cls(type=EventType.TIME_STOP, originator=orig)
        elif es == "time finish":
            return cls(type=EventType.TIME_FINISH, originator=orig)
        elif es.startswith("temp has reached"):
            return cls(type=EventType.TEMP_REACHED, originator=orig)
        else:
            raise ValueError(f"Unknown event: {event_string}")

    @staticmethod
    def is_event(message: str) -> bool:
        return message.startswith("event") or message.startswith("user changed")
